source/grid/manifold_lib.cc

```diff
@@ -361,80 +361,231 @@ get_tangent_vector (const Point<spacedim> &p1,
+      // is usually at most 8 points large.
+      for (unsigned int j = 0; j < n_unique_directions; ++j)
+        {
+          const double squared_distance = (directions[i] - directions[j]).norm_square();
+          if (!found_duplicate && squared_distance < 1e-28)
+            {
+              found_duplicate = true;
+              for (unsigned int row = 0; row < weights.size(0); ++row)
+                merged_weights[row][j] += weights[row][i];
+            }
+        }
+
+      if (found_duplicate == false)
+        {
+          merged_directions[n_unique_directions] = directions[i];
+          merged_distances[n_unique_directions] = distances[i];
+          for (unsigned int row = 0; row < weights.size(0); ++row)
+            merged_weights[row][n_unique_directions] = weights[row][i];
+
+          ++n_unique_directions;
+        }
+    }
+
+  // Note that we only use the n_unique_directions first entries in the ArrayView
+  const ArrayView<const Tensor<1,spacedim>> array_merged_directions = make_array_view(merged_directions.begin(),
+                                         merged_directions.begin()+n_unique_directions);
+  const ArrayView<const double> array_merged_distances = make_array_view(merged_distances.begin(),
+                                                         merged_distances.begin()+n_unique_directions);
+
+  for (unsigned int row=0; row<weights.size(0); ++row)
+    if (!accurate_point_was_found[row])
+      {
+        const ArrayView<const double> array_merged_weights (merged_weights[row].begin(),n_unique_directions);
+        new_points[row] = get_new_point(array_merged_directions,
+                                        array_merged_distances,
+                                        array_merged_weights,
+                                        new_points[row]);
+      }
+}
+
+
+
 template <int dim, int spacedim>
 Point<spacedim>
 SphericalManifold<dim,spacedim>::
 get_new_point (const ArrayView<const Point<spacedim>> &vertices,
                const ArrayView<const double>          &weights) const
 {
-  const unsigned int n_points = vertices.size();
+  // To avoid duplicating all of the logic in get_new_points, simply call it
+  // for one position.
+  const Table<2,double> table_weights(1,weights.size(),weights.begin());
+  Point<spacedim> new_point;
+  ArrayView<Point<spacedim>> new_points(&new_point,1);
+
+  get_new_points(vertices,table_weights,new_points);
+  return new_points[0];
+}
 
-  if (n_points == 2)
-    return get_intermediate_point(vertices[0], vertices[1], weights[1]);
 
-  const double tolerance = 1e-10;
-  const int max_iterations = 10;
 
+template <int dim, int spacedim>
+std::pair<double, Tensor<1,spacedim> >
+SphericalManifold<dim,spacedim>::
+guess_new_point(const ArrayView<const Tensor<1,spacedim>> &directions,
+                const ArrayView<const double> &distances,
+                const ArrayView<const double> &weights) const
+{
+  const double tolerance = 1e-10;
```

Comment:
We are using this value at several occasions in this class. I guess it makes sense to just turn it into a class variable.